synapseclient/models/dataset.py

```diff
@@ -403,15 +412,138 @@ def to_synapse_request(self):
         delete_none_keys(result)
         return result
 
-    # TODO: Implement this method
-    @staticmethod
-    async def snapshot_async(
+    def _append_entity_ref(self, entity_ref: EntityRef) -> None:
+        """Helper function to add an EntityRef to the items list of the dataset.
+        Will not add duplicates.
+
+        Arguments:
+            entity_ref: The EntityRef to add to the items list of the dataset.
+        """
+        if entity_ref not in self.items:
+            self.items.append(entity_ref)
+
+    @async_to_sync
+    async def add_item_async(
         self,
-        comment: str = None,
-        label: str = None,
-        include_activity: bool = True,
-        associate_activity_to_new_version: bool = True,
+        item: Union[EntityRef, "File", "Folder"],
         *,
         synapse_client: Optional[Synapse] = None,
-    ) -> Dict[str, Any]:
-        pass
+    ) -> None:
+        """Adds an item in the form of an EntityRef to the dataset.
+        For Folders, children are added recursively. Effect is not seen
+        until the dataset is stored.
+
+        Arguments:
+            item: Entity to add to the dataset. Must be an EntityRef, File, or Folder.
+            synapse_client: The Synapse client to use. Defaults to None.
+
+        Raises:
+            ValueError: If the item is not an EntityRef, File, or Folder
+        """
+        from synapseclient.models import File, Folder
+
+        client = Synapse.get_client(synapse_client=synapse_client)
+
+        if isinstance(item, EntityRef):
+            self._append_entity_ref(entity_ref=item)
+        elif isinstance(item, File):
+            file = File(
+                id=item.id, version_number=item.version_number, download_file=False
+            ).get()
+            self._append_entity_ref(
+                entity_ref=EntityRef(id=file.id, version=file.version_number)
+            )
```

Comment:
There is one improvement here where if `item.version_number` is filled in we can use that value, but if `item.version_number` isn't filled in then we'll need to get the `File` in order to populate the `EntityRef`